servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentSubscription.java

```diff
@@ -71,74 +74,59 @@ public static ConcurrentSubscription wrap(Subscription subscription) {
+        } else {
+            for (;;) {
+                final long prevPendingDemand = pendingDemand;
+                if (prevPendingDemand < 0 ||
+                        pendingDemandUpdater.compareAndSet(this, prevPendingDemand,
+                                prevPendingDemand + min(Long.MAX_VALUE - prevPendingDemand, n))) {
+                    break;
                 }
             }
-            if (n < 0) {
-                subscription.cancel();
-                return; // Don't set subscriptionLockOwner = 0 ... we don't want to request any more!
+        }
+        Throwable delayedCause = null;
+        for (;;) {
+            final long acquireId = acquirePendingReentrantLock(subscriptionLockUpdater, this);
+            if (acquireId == 0) {
+                break;
             }
+
             try {
-                subscription.request(n);
+                final long prevPendingDemand = pendingDemandUpdater.getAndSet(this, 0);
+                if (prevPendingDemand == CANCELLED) {
+                    subscription.cancel();
+                } else if (prevPendingDemand != 0) {
+                    subscription.request(prevPendingDemand);
+                }
+            } catch (Throwable cause) {
+                if (delayedCause == null) {
+                    delayedCause = cause;
+                }
             } finally {
-                subscriptionLockOwner = null;
+                if (releasePendingReentrantLock(subscriptionLockUpdater, acquireId, this)) {
+                    break;
+                }
             }
-            n = subscriptionRequestQueueUpdater.getAndSet(this, 0);
-        } while (n != 0);
+        }
+
+        if (delayedCause != null) {
+            throwException(delayedCause);
+        }
     }
 
     @Override
     public void cancel() {
-        // Set the queue to MIN_VALUE and this will be detected in request(n).
-        // We unconditionally set this value just in case there is re-entry with request(n) we will avoid calling
-        // the subscription's request(n) after cancel().
-        subscriptionRequestQueueUpdater.set(this, Long.MIN_VALUE);
-
-        final Thread currentThread = Thread.currentThread();
-        final Thread subscriptionLockOwner = this.subscriptionLockOwner;
-        if (subscriptionLockOwner == currentThread || subscriptionLockOwnerUpdater.compareAndSet(this, null,
-                currentThread)) {
+        pendingDemand = CANCELLED;
+        if (acquirePendingReentrantLock(subscriptionLockUpdater, this) != 0) {
             subscription.cancel();
+            // poison subscriptionLockUpdater
         }
     }
+
+    private static long mapInvalidRequestN(long n) {
+        // We map zero to a negative number because zero could later be overwritten by a subsequent legit value of
+        // n, and we want to ensure the invalid use gets propagated.
+        return n == CANCELLED ? CANCELLED + 1 : n == 0 ? -1 : n;
```

Comment:
nit: Imho chaining stuff like this is really not readable. Consider using a branch

Reply:
@Scottmitch ^^

Reply:
I prefer to keep the code as is as it is more compact and I don't think readability is dramatically different between the two options:

current option:
```
return n == CANCELLED ? CANCELLED + 1 : n == 0 ? -1 : n;
```

suggested option:
```
if (n == CANCELLED) {
  return CANCELLED + 1;
}
return n == 0 ? -1 : n;
```